File: tests/leetcode/x/python/0132.py
import sys


def min_cut(s):
    n = len(s)
    pal = [[False] * n for _ in range(n)]
    cuts = [0] * n
    for end in range(n):
        cuts[end] = end
        for start in range(end + 1):
            if s[start] == s[end] and (end - start <= 2 or pal[start + 1][end - 1]):
                pal[start][end] = True
                if start == 0:
                    cuts[end] = 0
                else:
                    cuts[end] = min(cuts[end], cuts[start - 1] + 1)
    return cuts[-1]


lines = sys.stdin.read().splitlines()
if lines:
    tc = int(lines[0])
    out = [str(min_cut(lines[i])) for i in range(1, tc + 1)]
    sys.stdout.write("\n\n".join(out))

File: tests/leetcode/x/ruby/0132.rb
def min_cut(s)
  n = s.length
  pal = Array.new(n) { Array.new(n, false) }
  cuts = Array.new(n, 0)
  (0...n).each do |ending|
    cuts[ending] = ending
    (0..ending).each do |start|
      if s[start] == s[ending] && (ending - start <= 2 || pal[start + 1][ending - 1])
        pal[start][ending] = true
        cuts[ending] = start.zero? ? 0 : [cuts[ending], cuts[start - 1] + 1].min
      end
    end
  end
  cuts[-1]
end

lines = STDIN.read.split(/\r?\n/)
exit if lines.empty? || lines[0].empty?
tc = lines[0].to_i
out = []
(1..tc).each { |i| out << min_cut(lines[i]).to_s }
STDOUT.write(out.join("\n\n"))
